File: mac-data.swift
import EventKit
import Foundation

let store = EKEventStore()
let semaphore = DispatchSemaphore(value: 0)
let args = CommandLine.arguments

guard args.count > 1 else { exit(1) }
let type = args[1]

if type == "calendar" {
    store.requestAccess(to: .event) { (granted, error) in
        if granted {
            let calendars = store.calendars(for: .event)
            let now = Date()
            let endOfDay = Calendar.current.date(bySettingHour: 23, minute: 59, second: 59, of: now)!
            let predicate = store.predicateForEvents(withStart: now, end: endOfDay, calendars: calendars)
            let events = store.events(matching: predicate)
            var out = [[String: String]]()
            let f = DateFormatter(); f.dateFormat="HH:mm"
            for e in events {
                out.append(["title": e.title ?? "", "start": f.string(from: e.startDate), "end": f.string(from: e.endDate)])
            }
            if let d = try? JSONSerialization.data(withJSONObject: out), let s = String(data: d, encoding: .utf8) { print(s) }
        }
        semaphore.signal()
    }
} else if type == "reminders" {
    store.requestAccess(to: .reminder) { (granted, error) in
        if granted {
            let predicate = store.predicateForIncompleteReminders(withDueDateStarting: nil, ending: nil, calendars: nil)
            store.fetchReminders(matching: predicate) { reminders in
                var out = [[String: String]]()
                for r in reminders ?? [] {
                    out.append(["name": r.title ?? "", "due": "none"])
                }
                if let d = try? JSONSerialization.data(withJSONObject: out), let s = String(data: d, encoding: .utf8) { print(s) }
                semaphore.signal()
            }
        } else { semaphore.signal() }
    }
}
semaphore.wait()

File: mac-data.js
ObjC.import('EventKit');
var store = $.EKEventStore.alloc.init;
var done = false;
var result = [];
var args = $.NSProcessInfo.processInfo.arguments;
var type = args.count > 3 ? args.objectAtIndex(3).js : "";

if (type === "calendar") {
    store.requestAccessToEntityTypeCompletion(0, function(granted, error) {
        if (granted) {
            var calendars = store.calendarsForEntityType(0);
            var cal = $.NSCalendar.currentCalendar;
            var startOfDay = cal.startOfDayForDate($.NSDate.date);
            var comps = $.NSDateComponents.alloc.init;
            comps.day = 1;
            var endOfDay = cal.dateByAddingComponentsToDateOptions(comps, startOfDay, 0);
            var predicate = store.predicateForEventsWithStartDateEndDateCalendars(startOfDay, endOfDay, calendars);
            var events = store.eventsMatchingPredicate(predicate);
            var f = $.NSDateFormatter.alloc.init;
            f.dateFormat = "HH:mm";
            for (var i = 0; i < events.count; i++) {
                var e = events.objectAtIndex(i);
                result.push({
                    title: e.title.js || "",
                    start: f.stringFromDate(e.startDate).js || "",
                    end: f.stringFromDate(e.endDate).js || ""
                });
            }
        }
        done = true;
    });
} else if (type === "reminders") {
    store.requestAccessToEntityTypeCompletion(1, function(granted, error) {
        if (granted) {
            var predicate = store.predicateForIncompleteRemindersWithDueDateStartingEndingCalendars(null, null, null);
            store.fetchRemindersMatchingPredicateCompletion(predicate, function(reminders) {
                if (reminders) {
                    for (var i = 0; i < reminders.count; i++) {
                        var r = reminders.objectAtIndex(i);
                        result.push({
                            name: r.title.js || "",
                            due: "none"
                        });
                    }
                }
                done = true;
            });
        } else {
            done = true;
        }
    });
} else {
    done = true;
}

while(!done) { $.NSRunLoop.currentRunLoop.runModeBeforeDate($.NSDefaultRunLoopMode, $.NSDate.dateWithTimeIntervalSinceNow(0.1)); }
JSON.stringify(result);
